### DISCUSSION TITLE
Building large datastrucutures (graphs) with FC

### DISCUSSION THREAD

**Discussion by Cypher1** [dubzzz/fast-check#4500] [Q&A]
Hi all,

I'm having some trouble building structures inside FC.
I have a Graph type with some methods (e.g. addNode(label), addEdge(from, to)) and I'm trying to build structures with a set of properties (like being forests of n-trees).

In fast check arbitrary code, this isn't 'hard' exactly, but the output is not particularly readable, as there's a lot of chaining and mapping.

I'd love to do something more monadic, unwrapping as I go, but the nesting gets very deep.

Am I missing some APIs for this in fastcheck?

Something like
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAMwMaKhEcDkCCGBnGAWiQAsBTJAa0wG4AoOpAG3zzgHEocwS4BvOnDgA7CABMyeAFxwCUYMIDmAbQC6cALxw19IWTGLJM5XIWKANLJjylqtZu2r6guEgjC5AVyQxoACigyHDF3JgBPOGAxGVMlS0Dg0IjhHBAyGOszAEp+AF8GIRAcSjIAOXEyPyiMm0UcgSEhGBJgPAA6UQl2sE88EiqxLN04fMLisgBRA0qENBAASWirWstfRZrs-hcmlvb9Q27e-uVZ9EXViEXVIZd8-LoYMLAyOAB5MBhgdzYtBrgQBTlLoAfhkwk8IAARmQoMMigAPIGSUEiCHQ2EuAHCKYHFHgqEwuE4eE45FgtGEuj3BCeYQ+L7COCKLg8AYbOIiVLpZZmSwQD4M6RvAXfBx8XL1FyBGCeKCMxBINpuDwwHDCGADLIuNpFMADTQAPhEZAA7hwWf0opYUmkslqhEqSDgFH5mYaFW1acAAI6eMgAQSgXDCfm2Htiij8WXMYb4WIAMmQlM0ZPzPt8dYCKnhLAjE8mSKmRR4dcSkXgJWHS3qbZJ3X9Gpy0u1dX5a+7maWSki21y7cNG9LZYzmQORnaq6RncJXVB3chPcIfX7A8HQ42hAuvb6A0GcCGw42FwoYGRDFA-HGFFIAAy54lSTudSRtJhJxTNOCEOAARgllivYR8w-Qs4AAJnveFgJTcD-0PRo+H+BRoNAtNBUzbFphzf5iRQot0xLBFSWwtxwBwLhfCgGQ-BwSxIRyDQjRwZQb3UDQtEhFj1AAMm4uBmJ-NiOOUQT4KESsN0aas-H2OtGK2STG1kltuBk6YOygLtJmmWcOmzZRZK41RLCffTDME64bkUxohzlJkMWsiUbnuZUCCZBwFzwVIwDfWduH6TAWkwSxAuAYK8n7RhvggN9XwgSM6hoIA

Thanks in advance :)

**Comment by Cypher1** [comment]
I had another go

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAMwMaKhEcDkCCGBnGAWiQAsBTJAa0wG4AoOpAG3zzgHEocwS4BvOnDgA7CABMyeAFxwCUYMIDmAbQC6cALxw19IWTGLJM5XIWKANLJjylqtZu2r6guEgjC5AVyQxoACigyHDF3JgBPOGAxGVMlS0Dg0IjhHBAyGOszAEp+AF8GIRAcSjIAOXEyPyiMm0UcgSEhGBJgPAA6UQl2sE88EiqxLN04fMLisgBRA0qENBAASWirWstfRZrs-hcmlvb9Q27e-uVZ9EXViEXVIZd8-LoYMLAyOAB5MBhgdzYtBrgQBTlLoAfhkwk8IAARmQoMMigAPIGSUEiCHQ2EuAHCKYHFHgqEwuE4eE45FgtGEuj3BCeYQ+L7COCKLg8AYbOIiVLpZZmSwQD4M6RvAXfBx8XL1FxuDzwHBQSHAaxysKkhx+ZkyTjcEg5DQAPkQSDatOAAEdPGQAIJQLhhPzbQ1tBQwMiGKB+PhYqQABksCKkzI6FXaTDISmacEIcAAjBLLJ6FAAZMOKZoyABMfuJyfDJAzozgNyE0oIcDlCqVUBV0x+cHVUE1LJ1mgNyGNwjNFuttvtjTL8sVXCrpPrWXMDr4-yTKbTcH5n2+bSxpLwWfhOdTebnIo8S+JK8sbnAcpwvgbdZwlkhuoNOGU3vUGi0kPv6gAZG+y8po4-n9-VLcNwuIEMCeFAjJtoEbhQGIHoOtUQhtiWMA4MIMADGODopGkMhId8KFoX42FkJhfadEYjomuaVo2jgdptrEih+GO-BYhus7zoKS6AsGa7sVunGLgiSJ4BKgF7mAfgelEljEbJwYSi2Wx9qG8DMg4whkAA7hwTYDLJXJFo0CDQHWqkiBUc4IBZXSSn2QiBkUJRIkRFRZA6BZCCBYGMsywxiUIbSkDgCj1i2DrloOyorqOEl+PskhKX8fYmVAZlkPACWRIyCV4HZ9kOVAe4lCOCWvpYZU-kZfaeY03ngUyGI1e5Qg3PcyFMg4DGpGAob1tqfiYC0mCWENwAjXkWQ3CWEChm0TAQExdQ0EAA

I think there's a repeated pattern of performing some side effecting code with an arbitrary generated from a previous value.

It would be really helpful to have something like

```
arbitrary1
    .with((value1) => makeSomeArbitrary2FromValue1))
    .map([value1, value2]) => {
          ... procedural code ...
    });
```

And maybe
```
arbitrary1
    .then((value1) => { some procedural code })
    .map(updatedValue1) => {
          ... imperative code ...
    });
```

I think that `then` (or some other name) represents that I want it to be side effecting, but still work for arbitraries.
Maybe returning the updated `value1` is acceptable, but ideally I don't want to have to.

**Comment by dubzzz** [comment]
Hello,

In the case of recursive structure, I'd go for `fc.letrec`. It will let you define recursive structures in an easier way.
Here is a pointer to the page describing it: https://fast-check.dev/docs/core-blocks/arbitraries/combiners/recursive-structure/#letrec

Otherwise, you can try `fc.memo`, but it's less recommended.

Or `chain` (as you already did) and in last resort `fc.gen` (last resort for `fc.gen` as it tends to provide bad shrinking, no replay...).

**Reply by dubzzz** [reply]
@Cypher1 did it solve your issue or help into solving it? If not, please let me know I can try to further investigate this precise case.
